plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -0,0 +1,860 @@
+		tmp.bits.First = (i == 0);
+		tmp.bits.Last = (i == bufsz - 1);
+		fu_bcm57xx_device_bar_write (self, FU_BCM57XX_BAR_DEVICE, REG_NVM_COMMAND, tmp.r32);
+		if (!fu_bcm57xx_device_nvram_wait_done (self, error)) {
+			g_prefix_error (error, "failed to read @0x%x: ", address);
+			return FALSE;
+		}
+		address += sizeof(guint32);
+		fu_device_set_progress_full (FU_DEVICE (self), i, bufsz);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_device_reset_ape (FuBcm57xxDevice *self, GError **error)
+{
+	/* halt */
+	RegAPEMode_t mode = { 0 };
+	mode.bits.Halt = 1;
+	mode.bits.FastBoot = 0;
+	fu_bcm57xx_device_bar_write (self, FU_BCM57XX_BAR_APE,
+				     REG_APE_MODE, mode.r32);
+
+	/* boot */
+	mode.bits.Halt = 0;
+	mode.bits.FastBoot = 0;
+	mode.bits.Reset = 1;
+	fu_bcm57xx_device_bar_write (self, FU_BCM57XX_BAR_APE,
+				     REG_APE_MODE, mode.r32);
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_device_detach (FuDevice *device, GError **error)
+{
+	/* unbind tg3 */
+	return fu_device_unbind_driver (device, error);
+}
+
+static gboolean
+fu_bcm57xx_device_attach (FuDevice *device, GError **error)
+{
+	/* bind tg3 */
+	return fu_device_bind_driver (device, "pci", "tg3", error);
+}
+
+static GBytes *
+fu_bcm57xx_device_dump_firmware (FuDevice *device, GError **error)
+{
+	FuBcm57xxDevice *self = FU_BCM57XX_DEVICE (device);
+	gsize bufsz_dwrds = fu_device_get_firmware_size_max (FU_DEVICE (self)) / sizeof(guint32);
+	g_autofree guint32 *buf_dwrds = g_new0 (guint32, bufsz_dwrds);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* not detached */
+	fu_bcm57xx_device_ensure_ethtool_iface (self);
+	if (self->ethtool_iface != NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "Not supported; detach required");
+		return FALSE;
+	}
+
+	/* read from hardware */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+	locker = fu_device_locker_new_full (self,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_enable,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_disable,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_bcm57xx_device_nvram_read (self, 0x0, buf_dwrds, bufsz_dwrds, error))
+		return NULL;
+	return g_bytes_new (buf_dwrds, bufsz_dwrds * sizeof(guint32));
+}
+
+static FuFirmware *
+fu_bcm57xx_device_read_firmware (FuDevice *device, GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_bcm57xx_firmware_new ();
+	g_autoptr(GBytes) fw = NULL;
+
+	/* read from hardware */
+	fw = fu_bcm57xx_device_dump_firmware (device, error);
+	if (fw == NULL)
+		return NULL;
+	if (!fu_firmware_parse (firmware, fw, FWUPD_INSTALL_FLAG_NONE, error))
+		return NULL;
+
+	/* remove images that will contain user-data */
+	if (!fu_firmware_remove_image_by_id (firmware, "info", error))
+		return NULL;
+	if (!fu_firmware_remove_image_by_id (firmware, "info2", error))
+		return NULL;
+	if (!fu_firmware_remove_image_by_id (firmware, "vpd", error))
+		return NULL;
```

Comment:
This causes the firmware image that is written back to hardware to have these sections erased.

Reply:
My mistake, thanks. Fixed.